server/parsec/asgi/rpc.py

```diff
@@ -837,16 +846,36 @@ async def __call__(self, scope, receive, send) -> None:
         else:
             self.client_ctx.logger.info("SSE session end")
 
+    @asynccontextmanager
+    async def listen_events_context(self) -> AsyncIterator[ApiEventListenOutcome]:
+        # This task is responsible for registering the client to the SSE stream and unregistering it
+        # when it finishes. It can either get cancelled internally (e.g. when the client gets frozen
+        # or revoked) or externally by the task group below (e.g. when the client gets disconnected).
+        async def listen_events(task_status: anyio.abc.TaskStatus[ApiEventListenOutcome]) -> None:
+            async with self.backend.events.sse_api_events_listen(
+                client_ctx=self.client_ctx, last_event_id=self.last_event_id
+            ) as outcome:
+                task_status.started(outcome)
+                await anyio.sleep_forever()
+
+        # Create a task group so that we can run the `sse_api_events_listen` context manager
+        # in parallel with the main task managing the SSE stream. It's important to run it
```

Comment:
So by "affect the main task" you mean `self.backend.events.sse_api_events_listen`'s cancel scope would cancel starlette's `StreamingResponse.stream_response`

Have you considered using the background parameter parameter instead ?
https://github.com/encode/starlette/blob/fa7b382a66cd99e3dc18f3baa44dae5ec68be76b/starlette/responses.py#L365

My guess is we didn't use it before because we need `initial_organization_config_event` in `_stream`, but it is only defined when calling `listen_events_context` which would be done in the background task...

However since we now need yet another task group (to isolate `_stream` and the background task, which depend on each other because we need share `initial_organization_config_event`...), it might be simpler to instead have background task pass `initial_organization_config_event` to `_stream` through a one-shot channel (so that they are isolated from the beginning, this would also remove the need to declare this custom `StreamingResponseMiddleware`).

(I leave it up to you to decide if it's worth it ^^)

Reply:
Wow this is some serious footgun xD